README.md
[![Build Status](https://travis-ci.org/n3okill/enfspatch.svg)](https://travis-ci.org/n3okill/enfspatch)
[![Build status](https://ci.appveyor.com/api/projects/status/524gen7vw7csbek8/branch/master?svg=true)](https://ci.appveyor.com/project/n3okill/enfspatch/branch/master)
[![Codacy Badge](https://api.codacy.com/project/badge/grade/e75d4c3ee1da4ff5a4d48f7dadb308bb)](https://www.codacy.com/app/n3okill/enfspatch)
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=64PYTCDH5UNZ6)

[![NPM](https://nodei.co/npm/enfspatch.png)](https://nodei.co/npm/enfspatch/)

enfspatch
=========
Additional methods and patches for node fs module

**enfs** stands for [E]asy [N]ode [fs]

This module is intended to work as a sub-module of [enfs](https://www.npmjs.com/package/enfs)


Description
-----------
This module will change some behaviors of fs module from node
such as creating a queue for opening files when limit is reached,
catching the error's and proceeding with the process when possible.
This module implements many methods through [enfsaddins](https://www.npmjs.com/package/enfsaddins)
  
Usage
-----
`enfspatch` is a drop-in replacement for native `fs` module, you just need to include
it instead of the native module.

Use this
```js
    var enfs = require("enfspatch");
```

instead of

```js
    var fs = require("fs"); //You don't need to do this anymore
```

and all the methods from native fs module are available

Errors
------
All the methods follows the node culture.
- Async: Every async method returns an Error in the first callback parameter
- Sync: Every sync method throws an Error.


Credit
------

This module is based on [graceful-fs](https://github.com/isaacs/node-graceful-fs)
- [Isaac Shlueter](https://github.com/isaacs)


License
-------

Creative Commons Attribution 4.0 International License

Copyright (c) 2016 Joao Parreira <joaofrparreira@gmail.com> [GitHub](https://github.com/n3okill)

This work is licensed under the Creative Commons Attribution 4.0 International License. 
To view a copy of this license, visit [CC-BY-4.0](http://creativecommons.org/licenses/by/4.0/).




CHANGELOG.md
1.0.0 / 2017-02-13
==================
  * dependencies update
  * rename async improvement
  * rename tests
  * general code improvements
  * major version
  
0.1.0 / 2016-04-11
==================
  * es6 code update
  * dependencies update

0.0.4 / 2016-02-20
==================
  * additions to .npmignore

0.0.3 / 2016-02-19
==================
  * license update
  
0.0.2 / 2016-02-19
==================
  * readme badges setup
  
0.0.1 / 2016-02-19
==================

  * first release


test/open.js
/**
 * @project enfspatch
 * @filename open.js
 * @description tests for enfspatch
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.1
 */

/* global describe, require, it, __filename */

"use strict";

const should = require("should");

describe("enfspatch > Open", function() {
    const fs = require("../");
    it("should open an existing file sync", function() {
        let fd;
        (function() {
            fd = fs.openSync(__filename, "r");
        }).should.not.throw();
        (function() {
            fs.closeSync(fd);
        }).should.not.throw();
    });
    it("should open an existing file async", function(done) {
        fs.open(__filename, "r", function(err, fd) {
            (err === null).should.be.equal(true);
            should.exist(fd);
            //(fd === undefined).should.be.equal(false);
            fs.close(fd, function(err) {
                (err === null).should.be.equal(true);
                done();
            });
        });
    });
    it("should fail to open non-existing file sync", function() {
        (function() {
            fs.openSync("invalid file path", "r");
        }).should.throw(Error, {code: "ENOENT", message: /no such file/});
    });
    it("should fail to open non-existing file async", function(done) {
        fs.open("invalid file path async", "r", function(err, fd) {
            should.not.exist(fd);
            //(fd === undefined).should.be.equal(true);
            err.should.be.instanceOf(Error);
            err.code.should.be.equal("ENOENT");
            //err.message.should.containEql("no such file");
            done();
        });
    });
});


test/stats.js
/**
 * @project enfspatch
 * @filename stats.js
 * @description tests for enfspatch
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.1
 */

/* global describe, require, it, __filename */

"use strict";

describe("enfspatch > Stats", function() {
    it("enfs should use the same stats constructor as fs module", function(done) {
        var fs, enfs;
        fs = require("fs");
        enfs = require("../");
        enfs.Stats.should.be.equal(fs.Stats);
        fs.stat(__filename, function(err, statFs) {
            (err === null).should.be.equal(true);
            statFs.should.be.instanceof(fs.Stats);
            enfs.stat(__filename, function(errEsnofs, statEsnofs) {
                (errEsnofs === null).should.be.equal(true);
                statEsnofs.should.be.instanceof(enfs.Stats);
                statEsnofs.should.be.instanceof(fs.Stats);
                statEsnofs.should.be.eql(statFs);
                done();
            });
        });
    });
});


test/addins.js
/**
 * @project enfspatch
 * @filename addins.js
 * @description tests for enfspatch
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.1
 */

/* global describe, require, it, __filename */

"use strict";


describe("enFsPatch-addins", function() {
    var fs = require("../");
    describe("> existStat", function() {
        describe("> async", function() {
            it("should test true for async __filename", function(done) {
                fs.existStat(__filename, function(err, result) {
                    result.should.be.equal(true);
                    done();
                });
            });
            it("should test false for async non-existent file", function(done) {
                fs.existStat("/not/existent/file", function(err, result) {
                    result.should.be.equal(false);
                    done();
                });
            });
        });
        describe("> sync", function() {
            it("should test true for sync __filename", function(done) {
                fs.existStatSync(__filename).should.be.equal(true);
                done();
            });
            it("should test false for sync non-existent file", function(done) {
                fs.existStatSync("/not/existent/file").should.be.equal(false);
                done();
            });
        });
    });
    describe("> existAccess", function() {
        describe("> async", function() {
            it("should test true for async __filename", function(done) {
                fs.existAccess(__filename, function(err, result) {
                    result.should.be.equal(true);
                    done();
                });
            });
            it("should test false for async non-existent file", function(done) {
                fs.existAccess("/not/existent/file", function(err, result) {
                    result.should.be.equal(false);
                    done();
                });
            });
        });
        describe("> sync", function() {
            it("should test true for sync __filename", function(done) {
                fs.existAccessSync(__filename).should.be.equal(true);
                done();
            });
            it("should test false for sync non-existent file", function(done) {
                fs.existAccessSync("/not/existent/file").should.be.equal(false);
                done();
            });
        });
    });
});


lib/addins.js
/**
 * @project enfspatch
 * @filename addins.js
 * @description Add functions to the fs module
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.0.1
 */


function addins(fs) {
    fs.existStat = function (path, callback) {
        fs.stat(path, function (err) {
            if (err) {
                return callback(null, false);
            }
            return callback(null, true);
        });
    };
    fs.existLStat = function (path, callback) {
        fs.lstat(path, function (err) {
            if (err) {
                return callback(null, false);
            }
            callback(null, true);
        });
    };
    fs.existFStat = function (fd, callback) {
        fs.fstat(fd, function (err) {
            if (err) {
                return callback(null, false);
            }
            callback(null, true);
        });
    };
    fs.existStatSync = function (path) {
        try {
            fs.statSync(path);
            return true;
        } catch (err) {
            return false;
        }
    };
    fs.existLStatSync = function(path) {
        try {
            fs.lstatSync(path);
            return true;
        } catch(err) {
            return false;
        }
    };
    fs.existSStatSync = function(fd) {
        try {
            fs.fstatSync(fd);
            return true;
        }catch(err) {
            return false;
        }
    };
    fs.existAccess = function (path, mode, callback) {
        if (typeof mode === "function") {
            callback = mode;
            mode = null;
        }
        fs.access(path, mode, function (err) {
            if (err) {
                return callback(null, false);
            }
            callback(null, true);
        });
    };
    fs.existAccessSync = function (path, mode) {
        try {
            fs.accessSync(path, mode);
            return true;
        } catch (err) {
            return false;
        }
    };
    fs.exists = (function (exists) {
        return function (path, callback) {
            if (exists) {
                exists(path, function (e) {
                    callback(null, e);
                });
            } else {
                fs.existStat(path, callback);
            }
        }
    })(fs.exists);

    fs.existsSync = fs.existsSync || fs.existStatSync;
}


module.exports = addins;


lib/polyfills.js
/**
 * @project enfspach
 * @filename polyfills.js
 * @description patch fs module based on Gracefull-fs module
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 1.0.0
 */

/*global require, global, process, module*/

"use strict";

const nodeConstants = require("constants");

function noop() {
}


function patchLutimes(fs) {
    if (nodeConstants.hasOwnProperty("O_SYMLINK")) {
        fs.lutimes = function (path, at, mt, cb) {
            fs.open(path, nodeConstants.O_SYMLINK, function (er, fd) {
                cb = cb || noop;
                if (er) {
                    return cb(er);
                }
                fs.futimes(fd, at, mt, function (er) {
                    fs.close(fd, function (er2) {
                        return cb(er || er2);
                    });
                });
            });
        };

        fs.lutimesSync = function (path, at, mt) {
            let ret;
            let fd = fs.openSync(path, nodeConstants.O_SYMLINK);

            let threw = true;
            try {
                ret = fs.futimesSync(fd, at, mt);
                threw = false;
            } finally {
                if (threw) {
                    try {
                        fs.closeSync(fd);
                    } catch (er) {
                        //this is intentional, we don't care if close throws an error because fd pointer
                        //may not exist at all
                    }
                } else {
                    fs.closeSync(fd);
                }
            }
            return ret;
        };
    } else {
        fs.lutimes = function (_a, _b, _c, cb) {
            process.nextTick(cb);
        };
        fs.lutimesSync = noop;
    }
}


// ENOSYS means that the fs doesn't support the op. Just ignore
// that, because it doesn't matter.
//
// if there's no getuid, or if getuid() is something other
// than 0, and the error is EINVAL or EPERM, then just ignore
// it.
//
// This specific case is a silent failure in cp, install, tar,
// and most other unix tools that manage permissions.
//
// When running as root, or if other types of errors are
// encountered, then it's strict.
function processUidErr(err) {
    return (!process.getuid || process.getuid() !== 0) && (err.code === "EINVAL" || err.code === "EPERM");
}

function chownErOk(err) {
    if(!err || err && err.code === "ENOSYS") {
        return true;
    }
    return processUidErr(err);
    /*if (!process.getuid || process.getuid() !== 0) {
        if (err.code === "EINVAL" || err.code === "EPERM") {
            return true;
        }
    }
    return false;*/
}


function chownFix(fs, orig) {
    if (!orig) {
        return orig;
    }
    return function (target, uid, gid, cb) {
        return orig.call(fs, target, uid, gid, function (err, res) {
            if (chownErOk(err)) {
                err = null;
            }
            cb(err, res);
        });
    };
}

function chownFixSync(fs, orig) {
    if (!orig) {
        return orig;
    }
    return function (target, uid, gid) {
        try {
            return orig.call(fs, target, uid, gid);
        } catch (err) {
            if (!chownErOk(err)) {
                throw err;
            }
        }
    };
}

function statFix(fs, orig) {
    if (!orig) {
        return orig;
    }
    // Older versions of Node erroneously returned signed integers for
    // uid + gid.
    return function (target, callback) {
        callback = callback || function(){};
        return orig.call(fs, target, function (er, stats) {
            if(stats) {
                if (stats.uid < 0) {
                    stats.uid += 0x100000000;
                }
                if (stats.gid < 0) {
                    stats.gid += 0x100000000;
                }
            }
            callback.apply(this, arguments);
        });
    };
}

function statFixSync(fs, orig) {
    if (!orig) {
        return orig;
    }
    // Older versions of Node erroneously returned signed integers for
    // uid + gid.
    return function (target) {
        let stats = orig.call(fs, target);
        if (stats.uid < 0) {
            stats.uid += 0x100000000;
        }
        if (stats.gid < 0) {
            stats.gid += 0x100000000;
        }
        return stats;
    };
}


module.exports = function patch(fs) {
    // (re-)implement some things that are known busted or missing.

    // lutimes implementation, or no-op
    if (!fs.lutimes) {
        patchLutimes(fs);
    }

    // https://github.com/isaacs/node-graceful-fs/issues/4
    // Chown should not fail on einval or eperm if non-root.
    // It should not fail on enosys ever, as this just indicates
    // that a fs doesn't support the intended operation.

    fs.chown = chownFix(fs, fs.chown);
    fs.fchown = chownFix(fs, fs.fchown);
    fs.lchown = chownFix(fs, fs.lchown);

    fs.chmod = chownFix(fs, fs.chmod);
    fs.fchmod = chownFix(fs, fs.fchmod);
    fs.lchmod = chownFix(fs, fs.lchmod);

    fs.chownSync = chownFixSync(fs, fs.chownSync);
    fs.fchownSync = chownFixSync(fs, fs.fchownSync);
    fs.lchownSync = chownFixSync(fs, fs.lchownSync);

    fs.chmodSync = chownFix(fs, fs.chmodSync);
    fs.fchmodSync = chownFix(fs, fs.fchmodSync);
    fs.lchmodSync = chownFix(fs, fs.lchmodSync);

    fs.stat = statFix(fs, fs.stat);
    fs.fstat = statFix(fs, fs.fstat);
    fs.lstat = statFix(fs, fs.lstat);

    fs.statSync = statFixSync(fs, fs.statSync);
    fs.fstatSync = statFixSync(fs, fs.fstatSync);
    fs.lstatSync = statFixSync(fs, fs.lstatSync);

    // if lchmod/lchown do not exist, then make them no-ops
    if (!fs.lchmod) {
        fs.lchmod = function (path, mode, cb) {
            process.nextTick(cb);
        };
        fs.lchmodSync = noop;
    }
    if (!fs.lchown) {
        fs.lchown = function (path, uid, gid, cb) {
            process.nextTick(cb);
        };
        fs.lchownSync = noop;
    }
};

lib/enfsPatch.js
/**
 * @project enfspatch
 * @filename enfspatch.js
 * @description Redefines open and close methods of fs module
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 1.0.0
 */

"use strict";


const enFs = require("fs");
const polyfills = require("./polyfills");
const enfsaddins = require("enfsaddins");
const platform = process.env.ENFS_FS_PLATFORM || process.platform;

let Queue = [], cwd = null;

function noop(){}

process.cwd = ((function (cwdFn) {
    return function () {
        if (!cwd) {
            cwd = cwdFn.call(process);
        }
        return cwd;
    };
})(process.cwd));

try {
    process.cwd();
} catch (er) {
    //Intentionally left blank
}


process.chdir = ((function (chdir) {
    return function (dir) {
        cwd = null;
        chdir.call(process, dir);
    };
})(process.chdir));


function addQueue(elem) {
    Queue.push(elem);
}

function processQueue() {
    let elem;
    elem = Queue.shift();
    if (elem) {
        elem[0].apply(null, elem[1]);
    }
}

// Always patch fs.close/closeSync, because we want to
// retry() whenever a close happens *anywhere* in the program.
enFs.close = ((function (close) {
    return function (fd, callback) {
        callback = callback || noop;
        return close.call(enFs, fd, function () {
            //retry even if close fails
            processQueue();
            callback.apply(this, arguments);
        });
    };
})(enFs.close));

enFs.closeSync = ((function (closeSync) {
    return function () {
        try {
            return closeSync.apply(enFs, arguments);
        } finally {
            //retry even if close fails
            processQueue();
        }
    };
})(enFs.closeSync));


function patch(fs) {
    // Everything that references the open() function needs to be in here
    polyfills(fs);
    enfsaddins(fs);

    fs.mockEnfs = patch;

    //This time is used in rename async and can be changed if needed (default: 1000ms)
    fs.renamingTime = 1000;

    // on Windows, A/V software can lock the directory, causing this
    // to fail with an EACCES or EPERM if the directory contains newly
    // created files.  Try again on failure, for up to 1 second.
    if (platform === "win32") {
        fs.rename = ((function (rename) {
            return function (from, to, callback) {
                callback = callback || noop;
                let stopTime = Date.now() + fs.renamingTime;
                let callAgainIn = 0;

                function cb(err) {
                    if (err && (err.code === "EACCES" || err.code === "EPERM") && (Date.now() < stopTime)) {
                        if (callAgainIn < 100) {
                            callAgainIn += 10;
                        }
                        setTimeout(function () {
                            fs.stat(to, function (statErr) {
                                if (statErr && statErr.code === "ENOENT") {
                                    return rename(from, to, cb);
                                }
                                callback(err);

                            });
                        }, callAgainIn);
                    } else {
                        callback(err);
                    }
                }

                return rename(from, to, cb);
            };
        })(fs.rename));

        fs.renameSync = ((function (rename) {
            return function (from, to) {
                let stopTime = Date.now() + 1000;

                function running() {
                    try {
                        return rename(from, to);
                    } catch (err) {
                        if (!((err.code === "EACCESS" || err.code === "EPERM") && (Date.now() < stopTime))) {
                            throw err;
                        } else {
                            try {
                                fs.statSync(to);
                            } catch (errStat) {
                                if (errStat.code === "ENOENT") {
                                    return running();
                                }
                                throw err;
                            }
                        }
                    }
                }
                return running();
            };
        })(fs.renameSync));
    }

    // if read() returns EAGAIN, then just try it again.
    fs.read = ((function (read) {
        return function (fd, buffer, offset, length, position, callback) {
            callback = callback || noop;
            let counter = 10;

            function fsR(fd, buffer, offset, length, position, callback) {
                return read(fd, buffer, offset, position, function (err) {
                    if (err && err.code === "EAGAIN" && --counter > 0) {
                        addQueue([fsR, [fs, buffer, offset, length, position, callback]]);
                    } else {
                        callback.apply(this, arguments);
                        processQueue();
                    }
                });
            }

            return fsR(fd, buffer, offset, length, position, callback);
        };
    })(fs.read));

    fs.readSync = ((function (readSync) {
        return function (fd, buffer, offset, length, position) {
            let counter = 10;

            function running() {
                try {
                    return readSync(fd, buffer, offset, length, position);
                } catch (err) {
                    if (err.code === "EAGAIN" && --counter > 0) {
                        return running();
                    }
                    throw err;
                }
            }

            return running();
        };
    })(fs.readSync));

    fs.readFile = ((function (readFile) {
        return function (path, options, callback) {
            callback = callback || noop;
            if (typeof options === "function") {
                callback = options;
                options = null;
            }

            function fsRF(path, options, callback) {
                return readFile(path, options, function (err) {
                    if (err && (err.code === "EMFILE" || err.code === "ENFILE")) {
                        addQueue([fsRF, [path, options, callback]]);
                    } else {
                        callback.apply(this, arguments);
                        processQueue();
                    }
                });
            }

            return fsRF(path, options, callback);
        };
    })(fs.readFile));

    fs.writeFile = ((function (writeFile) {
        return function (path, data, options, callback) {
            callback = callback || noop;
            if (typeof options === "function") {
                callback = options;
                options = null;
            }

            function fsWF(path, data, options, callback) {
                return writeFile(path, data, options, function (err) {
                    if (err && (err.code === "EMFILE" || err.code === "ENFILE")) {
                        addQueue([fsWF, [path, data, options, callback]]);
                    } else {
                        callback.apply(this, arguments);
                        processQueue();
                    }
                });
            }

            return fsWF(path, data, options, callback);
        };
    })(fs.writeFile));

    fs.appendFile = ((function (appendFile) {
        return function (path, data, options, callback) {
            callback = callback || noop;
            if (typeof options === "function") {
                callback = options;
                options = null;
            }

            function fsAF(path, data, options, callback) {
                return appendFile(path, data, options, function (err) {
                    if (err && (err.code === "EMFILE" || err.code === "ENFILE")) {
                        addQueue([fsAF, [path, data, options, callback]]);
                    } else {
                        callback.apply(this, arguments);
                        processQueue();
                    }
                });
            }

            return fsAF(path, data, options, callback);
        };
    })(fs.appendFile));

    fs.readdir = ((function (readdir) {
        return function (path, callback) {
            callback = callback || noop;

            function fsRd(path, callback) {
                return readdir(path, function (err, files) {
                    if (err && (err.code === "EMFILE" || err.code === "ENFILE")) {
                        addQueue([fsRd, [path, callback]]);
                    } else {
                        if (files && files.sort) {
                            //noinspection JSUnusedAssignment
                            files = files.sort();
                        }
                        callback.apply(this, arguments);
                        processQueue();
                    }
                });
            }

            return fsRd(path, callback);
        };
    })(fs.readdir));
    fs.readDir = fs.readdir;


    fs.open = ((function (open) {
        return function (path, flags, mode, callback) {
            callback = callback || noop;
            if (typeof mode === "function") {
                callback = mode;
                mode = null;
            }

            function fsO(path, flags, mode, callback) {
                return open(path, flags, mode, function (err) {
                    if (err && (err.code === "EMFILE" || err.code === "ENFILE")) {
                        addQueue([fsO, [path, flags, mode, callback]]);
                    } else {
                        //if (typeof callback === "function") {
                        callback.apply(this, arguments);
                        //}
                        processQueue();
                    }
                });
            }

            return fsO(path, flags, mode, callback);
        };
    })(fs.open));


    let OldReadStream = fs.ReadStream;

    function ReadStream() {
        if (this instanceof ReadStream) {
            return OldReadStream.apply(this, arguments);
        }
        return ReadStream.apply(Object.create(ReadStream.prototype), arguments);
    }

    ReadStream.prototype = Object.create(OldReadStream.prototype);
    ReadStream.prototype.open = function () {
        fs.open(this.path, this.flags, this.mode, (err, fd) => {
            if (err) {
                if (this.autoClose) {
                    this.destroy();
                }
                this.emit("error", err);
            } else {
                this.fd = fd;
                this.emit("open", this.fd);
                this.read();
            }
        });
    };
    function createReadStream(path, options) {
        return new ReadStream(path, options);
    }

    fs.createReadStream = createReadStream;
    fs.FileReadStream = ReadStream;
    fs.ReadStream = ReadStream;

    let OldWriteStream = fs.WriteStream;

    function WriteStream() {
        if (this instanceof WriteStream) {
            return OldWriteStream.apply(this, arguments);
        }
        return WriteStream.apply(Object.create(WriteStream.prototype), arguments);
    }

    WriteStream.prototype = Object.create(OldWriteStream.prototype);
    WriteStream.prototype.open = function () {
        fs.open(this.path, this.flags, this.mode, (err, fd) => {
            if (err) {
                this.destroy();
                this.emit("error", err);
            } else {
                this.fd = fd;
                this.emit("open", this.fd);
            }
        });
    };
    function createWriteStream(path, options) {
        return new WriteStream(path, options);
    }

    fs.createWriteStream = createWriteStream;
    fs.WriteStream = WriteStream;
    fs.FileWriteStream = WriteStream;

    return fs;
}


module.exports = patch(require("./fs"));
module.exports.enfsPatching = patch;